test/v2/controllers/RetrievePeriodSummaryControllerSpec.scala

```diff
@@ -101,106 +58,148 @@ class RetrievePeriodSummaryControllerSpec
-          }
-        }
-
-        val input = {
-          val errors = Seq(
-            (NinoFormatError, BAD_REQUEST),
-            (BusinessIdFormatError, BAD_REQUEST),
-            (PeriodIdFormatError, BAD_REQUEST),
-            (NotFoundError, NOT_FOUND),
-            (InternalError, INTERNAL_SERVER_ERROR),
-            (RuleIncorrectGovTestScenarioError, BAD_REQUEST)
-          )
-          val extraTysErrors = Seq(
-            (TaxYearFormatError, BAD_REQUEST),
-            (RuleTaxYearNotSupportedError, BAD_REQUEST),
-            (RuleTaxYearRangeInvalidError, BAD_REQUEST)
-          )
-          errors ++ extraTysErrors
-        }
-
-        input.foreach(args => (serviceErrors _).tupled(args))
+      "the service returns an error" in new Test {
+
+        MockRetrievePeriodSummaryRequestParser
+          .parse(rawData)
+          .returns(Right(requestData))
+
+        MockRetrievePeriodSummaryService
+          .retrieve(requestData)
+          .returns(Future.successful(Left(ErrorWrapper(correlationId, RuleTaxYearNotSupportedError))))
+
+        runErrorTest(RuleTaxYearNotSupportedError)
       }
     }
   }
 
+  trait Test extends ControllerTest {
+    val periodId: String                          = "2019-01-01_2020-01-01"
+    val rawData: RetrievePeriodSummaryRawData     = RetrievePeriodSummaryRawData(nino, businessId, periodId, None)
+    val requestData: RetrievePeriodSummaryRequest = RetrievePeriodSummaryRequest(Nino(nino), BusinessId(businessId), PeriodId(periodId), None)
+
+    val responseBody: RetrievePeriodSummaryResponse = RetrievePeriodSummaryResponse(
+      periodDates = PeriodDates("2019-01-01", "2020-01-01"),
+      periodIncome = None,
+      periodExpenses = None,
+      periodDisallowableExpenses = None
+    )
+
+    val testHateoasLink: Seq[Link] = Seq(
+      hateoas.Link(
+        href = s"/individuals/business/self-employment/$nino/$businessId/period/$periodId",
+        method = PUT,
+        rel = "amend-self-employment-period-summary"
+      ),
+      hateoas.Link(href = s"/individuals/business/self-employment/$nino/$businessId/period/$periodId", method = GET, rel = "self"),
+      hateoas.Link(
+        href = s"/individuals/business/self-employment/$nino/$businessId/period/$periodId",
+        method = GET,
+        rel = "list-self-employment-period-summaries"
+      )
+    )
+
+    val responseJson: JsValue = Json.parse(
+      s"""
+         |{"periodDates":{"periodStartDate":"2019-01-01","periodEndDate":"2020-01-01"},"links":[{"href":"/individuals/business/self-employment/AA123456A/XAIS12345678910/period/2019-01-01_2020-01-01","method":"PUT","rel":"amend-self-employment-period-summary"},{"href":"/individuals/business/self-employment/AA123456A/XAIS12345678910/period/2019-01-01_2020-01-01","method":"GET","rel":"self"},{"href":"/individuals/business/self-employment/AA123456A/XAIS12345678910/period/2019-01-01_2020-01-01","method":"GET","rel":"list-self-employment-period-summaries"}]}
```

Comment:
can format this as well,

plus `hateoas.Link` -> `Link` and make vals + traits private

Reply:
(if a val is inside a trait then can leave it as public